onnxruntime/core/graph/contrib_ops/onnx_function_util.cc

```diff
@@ -102,8 +102,61 @@ bool BuildFunctionProto(FunctionProto& functionProto, const OpSchema& schema,
                         const std::vector<FunctionBodyHelper::NodeDef>& node_defs,
                         const std::vector<OperatorSetIdProto>& relied_opsets) {
   BuildNodes(functionProto, node_defs);
-  schema.BuildFunction(functionProto, relied_opsets);
+  for (auto& relied_opset : relied_opsets) {
+    *(functionProto.mutable_opset_import()->Add()) = relied_opset;
+  }
+  schema.BuildFunction(functionProto);
   return true;
 }
 
+bool IsFunctionOpsetCompatible(
+    const ISchemaRegistry* schema_registry,
+    const FunctionProto& func_proto,
+    const std::unordered_map<std::string, int>& graph_imports) {
+  std::unordered_map<std::string, int> func_imports;
+  for (const auto& fn_import : func_proto.opset_import()) {
+    func_imports.insert({fn_import.domain(), static_cast<int>(fn_import.version())});
+  }
+
+  auto get_version = [](const std::string& domain, const std::unordered_map<std::string, int>& opset_imports) {
+    auto it = opset_imports.find(domain);
+    if (it == opset_imports.end()) {
+      return -1;
+    }
+    return it->second;
+  };
+
+  for (const auto& node : func_proto.node()) {
+    auto func_opset_version = get_version(node.domain(), func_imports);
+    auto model_opset_version = get_version(node.domain(), graph_imports);
+
+    if (func_opset_version == -1 && model_opset_version == -1) {
+      // No opset import for domain in graph and function imports.
+      return false;
+    } else if (func_opset_version == model_opset_version || model_opset_version == -1 || func_opset_version == -1) {
+      // If the opset versions are same or opset import for node.domain is present in one of the imports return true.
+      // ONNX allows model_opset_version == -1 but not func_opset_version == -1
+      // We relax the check in ORT to support pre-existing models.
```

Comment:
This discrepancy should eventually be resolved, right? If so, may be worth adding a TODO comment to that effect.

Reply:
as per offline discussion removing this function altogether.